tests/model/architecture.test.js

```diff
@@ -0,0 +1,190 @@
+        let a = new Architecture();
+
+        a.xSpan = -14;
+        a.ySpan = -9876543;
+        expect(a.determineParams()).toBe(false);
+
+        expect(a.hasParams).toBe(false);
+    });
+});
+
+describe('validation', () => {
+    describe('valid', () => {
+        test('with params', () => {
+            let a = new Architecture('arch', [topLayer, botLayer], 1000, 1000);
+
+            expect(a.validate()).toBe(true);
+        });
+
+        test('without params', () => {
+            let a = new Architecture('arch', [topLayer, botLayer]);
+
+            expect(a.validate()).toBe(true);
+        });
+    });
+
+    describe('invalid', () => {
+        test('all fields', () => {
+            let a = new Architecture();
+
+            expect(a.validate()).toBe(false);
+        });
+
+        test('name', () => {
+            let a = new Architecture(Validation.DEFAULT_STR_VALUE, [topLayer, botLayer]);
+
+            expect(a.validate()).toBe(false);
+        });
+
+        describe('layers', () => {
+            test('empty', () => {
+                let a = new Architecture('arch', []);
+
+                expect(a.validate()).toBe(false);
+            });
+
+            test('invalid component', () => {
+                let a = new Architecture('arch', [new Layer('bad-layer', 'unique-bad-layer-id', [new Component()],
+                        [connection1]), botLayer]);
+
+                expect(a.validate()).toBe(false);
+            });
+
+            test('invalid connection', () => {
+                let a = new Architecture('arch', [topLayer, new Layer('bad-layer', 'unique-bad-layer-id', [component1],
+                        [new Connection()])]);
+
+                expect(a.validate()).toBe(false);
+            });
+        });
+
+        describe('locations', () => {
+            test('Component Features', () => {
+                let a = new Architecture('tiny-arch', [topLayer, botLayer], 5, 5);
+
+                expect(a.validate()).toBe(false);
+            });
+
+            test('Connection Segments', () => {
+                let port1_1 = new Port('port', new Coord(1, 1));
+                let comp1 = new Component('tiny-component-1', 'unique-tiny-component-id-1', 1, 1, 'entity', [port0_0]);
+                let comp2 = new Component('tiny-component-2', 'unique-tiny-component-id-2', 1, 1, 'entity', [port1_1]);
+                let source = new Terminal(comp1, port0_0);
+                let sink = new Terminal(comp2, port1_1);
+                let conn = new Connection('connection', 'unique-connection-id', source, [sink],
+                        [new ConnectionSegment('con-seg-name', 'unique-con-seg-id', 5, 10, new Coord(0, 0),
+                        new Coord(50, 50))]);
+                let a = new Architecture('tiny-arch', [topLayer, new Layer('layer', 'unique-layer-id', [comp1, comp2],
+                        [conn])], 5, 5);
+
+                expect(a.validate()).toBe(false);
```

Comment:
I'm a bit confused about this test and the one above, why should they be false? Is it because one of the locations is outside the bounds of the device?

Reply:
Yes the one above tests a component that is too large for the Architecture size. This one tests connections that have sink- / source-points outside of the Architecture.